src/arch/cuda/hybridAPI/hapi_impl.cpp

```diff
@@ -131,28 +136,13 @@ void hapiInit(char** argv) {
 
     hapiMapping(argv); // Perform PE-device mapping
 
+#ifndef HAPI_CUDA_CALLBACK
     // Register polling function to be invoked at every scheduler loop
     CcdCallOnConditionKeep(CcdSCHEDLOOP, hapiPollEvents, NULL);
-
-    CmiNodeBarrier(); // Ensure PE-device mappings are complete within a logical node
-
-    if (CmiMyRank() == 0) {
-      shmCreate(); // Create a per-host shared memory region
-    }
-
-    CmiNodeBarrier(); // Ensure shared memory regions have been created
-
-    ipcHandleCreate(); // Create CUDA IPC handles
+#endif
   }
 
-  // Ensure CUDA IPC handles are available for all processes
-  // Note: Causes a hang when this barrier is placed after CPU topology initialization
-  // FIXME: This only needs to be a host-wide synchronization
-  CmiBarrier();
-
-  if (CmiMyRank() == 0) {
-    ipcHandleOpen(); // Open CUDA IPC handles for accessing other processes' device memory
-  }
```

Comment:
Moved to `shmInit`, so that these won't be executed when the user passes `+gpunoshm`.